crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+    // Restore the trailing slash if there was one (camino normalizes these away),
+    // as this is how zip archives distinguish between files and directories.
+    if path.as_str().ends_with('/') {
+        normalized.push('/');
+    }
+
+    NormalizedVendoredPath(normalized)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+    use std::io::Write;
+
+    use once_cell::sync::Lazy;
+    use tempfile::TempDir;
+    use zip::{write::FileOptions, CompressionMethod, ZipWriter};
+
+    use super::*;
+
+    const FUNCTOOLS_CONTENTS: &str = "def update_wrapper(): ...";
+    const ASYNCIO_TASKS_CONTENTS: &str = "class Task: ...";
+
+    static MOCK_ZIP_ARCHIVE: Lazy<Arc<[u8]>> = Lazy::new(|| {
+        let td = TempDir::new().unwrap();
+        let typeshed_path = td.path().join("typeshed");
+        let typeshed = fs::File::create(&typeshed_path).unwrap();
+
+        let options = FileOptions::default()
+            .compression_method(CompressionMethod::Zstd)
+            .unix_permissions(0o644);
+
+        let mut archive = ZipWriter::new(typeshed);
+
+        archive.add_directory("stdlib/", options).unwrap();
+        archive.start_file("stdlib/functools.pyi", options).unwrap();
+        archive.write_all(FUNCTOOLS_CONTENTS.as_bytes()).unwrap();
+
+        archive.add_directory("stdlib/asyncio/", options).unwrap();
+        archive
+            .start_file("stdlib/asyncio/tasks.pyi", options)
+            .unwrap();
+        archive
+            .write_all(ASYNCIO_TASKS_CONTENTS.as_bytes())
+            .unwrap();
+
+        archive.finish().unwrap();
+        Arc::from(fs::read(typeshed_path).unwrap().into_boxed_slice())
+    });
+
+    fn mock_typeshed() -> VendoredFileSystem {
+        VendoredFileSystem::new(Arc::clone(&MOCK_ZIP_ARCHIVE)).unwrap()
+    }
+
+    macro_rules! directory_tests {
+        ($($name:ident: $dirname:literal,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                let mock_typeshed = mock_typeshed();
+
+                let path = FileSystemPath::new($dirname);
+
+                assert!(mock_typeshed.exists(path));
+                assert!(mock_typeshed.is_directory(path));
+                assert!(!mock_typeshed.is_file(path));
+
+                let path_metadata = mock_typeshed.metadata(path).unwrap();
+                assert!(path_metadata.file_type().is_directory());
+                assert!(path_metadata.permissions().is_some());
+                assert_eq!(path_metadata.revision(), FileRevision::zero());
+            }
+        )*
+        }
+    }
+
+    directory_tests! {
+        stdlib_dir_no_trailing_slash: "stdlib",
+        stdlib_dir_trailing_slash: "stdlib/",
+        asyncio_dir_no_trailing_slash: "stdlib/asyncio",
+        asyncio_dir_trailing_slash: "stdlib/asyncio/",
+        stdlib_dir_parent_components: "stdlib/asyncio/../../stdlib",
+        asyncio_dir_odd_components: "./stdlib/asyncio/../asyncio/",
+    }
```

Comment:
I think we don't need to use macros here. They safe very little code but break IDE features. For example, I can't jump or run the test manually anymore. Rustfmt breaks down, and auto completion often doesn't work. They're also hard to read. 

```suggestion
    fn assert_directory(directory_name: &str) {
                let mock_typeshed = mock_typeshed();

                let path = FileSystemPath::new(directory_name);

                assert!(mock_typeshed.exists(path));
                assert!(mock_typeshed.is_directory(path));
                assert!(!mock_typeshed.is_file(path));

                let path_metadata = mock_typeshed.metadata(path).unwrap();
                assert!(path_metadata.file_type().is_directory());
                assert!(path_metadata.permissions().is_some());
                assert_eq!(path_metadata.revision(), FileRevision::zero());
    }
    
    #[test]
    fn stdlib_dir_no_trailing_slash() {
      assert_directory("stdlib");
    }
```

Same for the other tests

Reply:
Duh. I felt so clever about this... How did I forget about this much simpler solution 😑

(though FWIW, it didn't seem to break any IDE features for me on VSCode — I was still able to run the test manually locally)